perception/navigator_vision/nodes/cvod/cvod.py

```diff
@@ -0,0 +1,263 @@
+import os
+import cv2
+import numpy as np
+import sys
+import rospy
+import tensorflow
+from cv_bridge import CvBridge, CvBridgeError
+import image_geometry
+import mil_tools
+from sensor_msgs.msg import CameraInfo, Image
+import tf
+from mil_msgs.srv import ObjectDBQuery, ObjectDBQueryResponse
+from mil_msgs.msg import PerceptionObject
+
+sys.path.append("..")
+
+from cvod_utils import label_map_util
+from cvod_utils import visualization_utils as vis_util
+
+
+class CVOD(object):
+    '''
+    The Computer Vision based Object Detection is a trained neural network designed to detect buoys and totems in real time using cameras mounted on Navigator.
+    By subscribing to the camera outputs through ROS, we are able to scan the video feeds for objects, labelling them and giving the coordinates of the objects relative to the boat.
+    This is used in conjunction with the LIDAR to construct and occupancy grid that allows Navigator to autonomously avoid obstacles in real time.
+
+    The network was trained using the Tensorflow library and through the process of transfer learning. A pretrained COCO model was loaded into the trainer, with most of the layers 'frozen'.
+    These layers would not be changed during training. The last few layers of the network were then trained using our dataset of 400 images.
+
+    @TODO:
+
+    - Create a map to store ID from database and associated confidence level for each object. If new confidence is greater than or equal to previous confidence, update
+    - Find a way to weight updates to Ogrid, not just labels.
+    '''
+
+    def __init__(self):
+        rospy.init_node('active_detection')
+        self.tf_listener = tf.TransformListener()
+        ################ Init Database Services ###############################
+
+        rospy.wait_for_service('/database/requests')
+        self.DBQuery_proxy = rospy.ServiceProxy('/database/requests', ObjectDBQuery)
+
+        ################ Init Publishers and Bridge ###########################
+        # Image publisher for debugging, publishes images with bounding boxes
+        # and labels drawn around detected objects.
+
+        # Debug image for verification
+        self.image_pub = rospy.Publisher(
+            "/cvod/debug_images", Image, queue_size=1)
+
+        '''
+		CVBridge allows us to convert openCV images into ROS msgs and vice versa.
+		It is the active link between publishers, subscribers and the operations performed between.
+        '''
+        self.bridge = CvBridge()
+
+        ################ Init Subscribers #####################################
+        # Used to keep track of when we should update PCODAR.
+        self.see_frame_counter = 0
+        self.left_frame_counter = 0
+
+        # Image subscriber for seecam, these images will be processed by the
+        # network.
+        self.see_sub = mil_tools.Image_Subscriber(
```

Comment:
It would save code and make program more adaptable by just taking a list of camera topics as a ROS param instead of hard coding these two.